packages/client/src/clients/in-app-messages/socket-driver.ts

```diff
@@ -0,0 +1,187 @@
+    const referenceId = iamClient.referenceId;
+    const params = iamClient.defaultOptions;
+
+    // Ensure a live socket connection if not yet connected.
+    if (!this.socket.isConnected()) {
+      this.socket.connect();
+    }
+
+    // If a new in-app message client joins, or has updated query params then
+    // track the updated params and (re)join with the latest query params.
+    // Note, each time we send combined params of all in-app message clients that
+    // have subscribed for a given message type (and an in-app channel and a
+    // user), grouped by client's reference id.
+    if (!this.params[topic]) {
+      this.params[topic] = {};
+    }
+
+    const maybeParams = this.params[topic][referenceId];
+    const hasNewOrUpdatedParams =
+      !maybeParams || JSON.stringify(maybeParams) !== JSON.stringify(params);
+
+    if (hasNewOrUpdatedParams) {
+      // Tracks all subscribed client's params by its reference id and by topic.
+      this.params[topic] = { ...this.params[topic], [referenceId]: params };
+    }
+
+    if (!this.channels[topic] || hasNewOrUpdatedParams) {
+      const newChannel = this.socket.channel(topic, this.params[topic]);
+      for (const eventType of SOCKET_EVENT_TYPES) {
+        newChannel.on(eventType, (payload) => this.setInbox(payload));
+      }
+      // Tracks live channels by channel topic.
+      this.channels[topic] = newChannel;
+    }
+
+    const channel = this.channels[topic];
+
+    // Join the channel if not already joined or joining or leaving.
+    if (["closed", "errored"].includes(channel.state)) {
+      channel.join();
+    }
+
+    // Let the in-app message client subscribe to the "inbox", so it can be
+    // notified when there's a new socket event that is relevant for the client.
+    const unsub = this.inbox.subscribe(() => {
+      const payload = this.inbox.state[referenceId];
+      if (!payload) return;
+
+      iamClient.handleSocketEvent(payload);
+    });
+
+    return unsub;
+  }
+
+  leave(iamClient: InAppMessagesClient) {
+    if (iamClient.unsub) {
+      iamClient.unsub();
+    }
+
+    const topic = iamClient.socketChannelTopic();
+    const referenceId = iamClient.referenceId;
+
+    const partitionedParams = { ...this.params };
+    const paramsForTopic = partitionedParams[topic] || {};
+    const paramsForReferenceClient = paramsForTopic[referenceId];
+
+    if (paramsForReferenceClient) {
+      delete paramsForTopic[referenceId];
+    }
+
+    const channels = { ...this.channels };
+    const channelForTopic = channels[topic];
+    if (channelForTopic && Object.keys(paramsForTopic).length === 0) {
+      for (const eventType of SOCKET_EVENT_TYPES) {
+        channelForTopic.off(eventType);
+      }
+      channelForTopic.leave();
+      delete channels[topic];
+    }
+
+    this.params = partitionedParams;
+    this.channels = channels;
+  }
+
+  private setInbox(payload: WithAttn<SocketEventPayload>) {
+    const { attn, ...rest } = payload;
```

Comment:
`attn` field is provided by the socket API to indicate which in-app message clients should be notified of the incoming socket event based on respective query params. 
https://github.com/knocklabs/switchboard/pull/2963/files#diff-aa5f2245deb4ebc08de405e1f9cac816d0e942bcfffbfcd1693e8d3594d24427R137